### PR TITLE
Add better error reporting from the backends

### PR DESCRIPTION
**Pull request by abergeron**
This allows the backend to clarify errors by adding error messages that include relevant information.

This should make it much easier to track down problems reported by users and even allows a trace of all the errors encountered in debug mode.

The code depends on the disk_cache PR since that modifies a lot of the error handling and I didn't want to do the work twice.  I will rebase once that is in.

### CODE REVIEW

**Comment by abergeron** [comment]
jenkins test this please

**Comment by nouiz** [comment]
I have many tests failing:

libgpuarray/Build$ DEVICE=cuda0 make test
Running tests...
Test project /u/bastienf/repos/libgpuarray/Build
      Start  1: test_types
 1/11 Test  #1: test_types .......................   Passed    0.01 sec
      Start  2: test_util
 2/11 Test  #2: test_util ........................   Passed    0.01 sec
      Start  3: test_util_integerfactoring
 3/11 Test  #3: test_util_integerfactoring .......   Passed    0.49 sec
      Start  4: test_reduction
 4/11 Test  #4: test_reduction ...................***Failed    1.13 sec
      Start  5: test_array
 5/11 Test  #5: test_array .......................***Failed    0.59 sec
      Start  6: test_blas
 6/11 Test  #6: test_blas ........................***Failed    0.28 sec
      Start  7: test_elemwise
 7/11 Test  #7: test_elemwise ....................***Failed    3.48 sec
      Start  8: test_error
 8/11 Test  #8: test_error .......................   Passed    0.01 sec
      Start  9: test_buffer
 9/11 Test  #9: test_buffer ......................***Failed    1.71 sec
      Start 10: test_buffer_collectives
10/11 Test #10: test_buffer_collectives ..........***Failed    2.91 sec
      Start 11: test_collectives
11/11 Test #11: test_collectives .................***Failed    1.29 sec

36% tests passed, 7 tests failed out of 11

Total Test time (real) =  11.96 sec

The following tests FAILED:
	  4 - test_reduction (Failed)
	  5 - test_array (Failed)
	  6 - test_blas (Failed)
	  7 - test_elemwise (Failed)
	  9 - test_buffer (Failed)
	 10 - test_buffer_collectives (Failed)
	 11 - test_collectives (Failed)
Errors while running CTest
Makefile:84: recipe for target 'test' failed
make: *** [test] Error 8

and jenkins fail

**Comment by nouiz** [comment]
I double checked and the master "make test" tests pass.

**Comment by abergeron** [comment]
This includes a large portion of the cache code, which might be a broken version.  Once that part is in, I'll rebase this and we will see if it still fails.

**Comment by abergeron** [comment]
It appears jenkins had no GPU.

jenkins test this please

**Comment by nouiz** [comment]
Can you confirm it was rebased and adjusted on the master since the merge of the cache PR?

jenkins had no GPU.

jenkins test this again

**Comment by abergeron** [comment]
Yes, I did a rebase and it should be ok.  However since this is the third time jenkins appears to not have a GPU for this, maybe something is wrong with the code itself.

**Comment by nouiz** [comment]
jenkins test this

File: src/private.h [hunk 1]
```diff
@@ -76,13 +77,12 @@ typedef struct _partial_gpucomm {
 struct _gpuarray_buffer_ops {
   int (*get_platform_count)(unsigned int* platcount);
   int (*get_device_count)(unsigned int platform, unsigned int* devcount);
-  gpucontext *(*buffer_init)(int dev, int flags, int *ret);
+  gpucontext *(*buffer_init)(int dev, int flags);
   void (*buffer_deinit)(gpucontext *ctx);
-  gpudata *(*buffer_alloc)(gpucontext *ctx, size_t sz, void *data, int flags,
-                           int *ret);
+  gpudata *(*buffer_alloc)(gpucontext *ctx, size_t sz, void *data, int flags);
```
Comment (nouiz) [inline src/private.h]: interface change that should request a version bump?

**Comment by slefrancois** [comment]
jenkins test this

**Comment by slefrancois** [comment]
I rebooted the VMs and nvidia-smi shows GPUs are OK. It's probably a problem with the PR.

File: src/private.h (hunk 1 above)
Comment (abergeron) [inline src/private.h]: No, this is private.

**Comment by abergeron** [comment]
I should have fixed the problem jenkins was having.

**Comment by abergeron** [comment]
jenkins test this please

**Comment by slefrancois** [comment]
jenkins test this

**Comment by slefrancois** [comment]
jenkins test this

**Comment by lamblin** [comment]
The tests for libgpuarray (C) and pygpu pass locally for me.
I did not try theano/libgpuarray yet.

**Comment by lamblin** [comment]
Looks good to me, and @abergeron quickly checked my changes.
@nouiz I'll let you finish and merge.

**Comment by nouiz** [comment]
While running pygpu tests with OpenCL and no blas, I get this error:

"Collectives operations not supported on OpenCL"

This is the wrong error. Not sure how to fix that after a quick investigation.

**Comment by lamblin** [comment]
Right, I did not test with OpenCL... I'll do that.

**Comment by nouiz** [comment]
I pushed fix for that.

**Comment by nouiz** [comment]
Now opencl tests only fail due to elemwise in float16 that don't give exactly the same results.

**Comment by juancamilog** [comment]
build fails on Windows 10 + VS2015:

1. name is undeclared dyn_load.c lines 38 and 40
2. getting a syntax error on error.c line 8, before "}":  "expected an exprerssion"
3. syntax error at disk.c line 348 : "missing ';' before ':' "
4. same as 3. but in gpuarray_elemwise.c

**Comment by nouiz** [comment]
We merged a fix: https://github.com/Theano/libgpuarray/pull/415#pullrequestreview-33820720

Can you update libgpuarray and try again?

**Pull request merged**